examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,763 @@
+    auto worker_function = [&](tokens_pair & tp) {
+        printf("Evaluating prompt: \"%s\" - \"%s\" (%ld tokens)\n", tp.positive.c_str(), tp.negative.c_str(), tp.max_seq_len);
+        // TODO so how do we deal with this?
+        // TODO we only have one cb_data object that everything gets passed to. so we need to be able to write to a different object per thread
+        // TODO but there's only one cb_eval function used as callback by the model... help wanted
+    };
+    printf("Batching prompts...\n");
+    for (int i = 0; i < batched_prompts.size(); ++i) {
+        for (int j = 0; j < batched_prompts[i].size(); ++j) {
+            threads.emplace_back(worker_function, batched_prompts[i][j]);
+        }
+        for (auto & th : threads) th.join();
+        
+        // reload model for next batch
+        llama_free(ctx);
+        llama_free_model(model);
+        std::tie(model, ctx) = llama_init_from_gpt_params(params);
+    }
+    printf("Done with batching prompts.\n");
+    */
+
+    int token_ct = 0;
+
+    for(size_t i = 0; i < cparams.positive_entries.size(); ++i) {
+        std::string positive_prompt = cparams.positive_entries[i];
+        std::string negative_prompt = cparams.negative_entries[i];
+        std::vector<llama_token> tokens_pos = ::llama_tokenize(ctx, positive_prompt, add_bos);
+        std::vector<llama_token> tokens_neg = ::llama_tokenize(ctx, negative_prompt, add_bos);
+        size_t max_seq_len = std::max(tokens_pos.size(), tokens_neg.size());
+        padding_seq(ctx, tokens_pos, max_seq_len);
+        padding_seq(ctx, tokens_neg, max_seq_len);
+        cb_data.n_tokens = max_seq_len;
+
+        // need to reload the model so it doesn't run out of context
+        // this should scale with -c option passed by main
+        token_ct += 2 * max_seq_len;
+        if (token_ct > n_ctx || cparams.always_reload) {
+            //break;
+            llama_free(ctx);
+            llama_free_model(model);
+            std::tie(model, ctx) = llama_init_from_gpt_params(params);
+            token_ct = 2 * max_seq_len;
+        }
+        if (token_ct > n_ctx) {
+            fprintf(stderr, "context size exceeded on iteration %zu\n", i);
+            break;
+        }
+
+        printf("Evaluating prompt: \"%s\" - \"%s\" (%ld tokens)\n", positive_prompt.c_str(), negative_prompt.c_str(), max_seq_len);
+
+        cb_data.is_eval_pos = true;
+        get_hidden_layers(ctx, tokens_pos);
+        cb_data.is_eval_pos = false;
+        get_hidden_layers(ctx, tokens_neg);
+
+        // TODO check whether the same tokens correspond to zero rows because we don't seem to be getting many zero rows anymore
+        // we get a lot of zero rows for the first few prompts and then they drop off
+        // likewise most of the zero rows are in the first few layers for each prompt
+
+        calc_diff(cb_data);
+
+        // reset for next iteration
+        cb_data.v_pos.clear();
+        cb_data.v_neg.clear();
```

Comment:
Small note here: Because these vectors contains pointers, it's better to free them instead of just clear (that will cause memory leak). To free the pointers, you can do:

```
for (auto ptr : cb_data.v_pos) free(ptr);
```

It's not something too important though, we will have a better way to do when we move these vectors to tensor.